datafusion/optimizer/src/simplify_expressions/simplify_sql_literal.rs

```diff
@@ -0,0 +1,232 @@
+    plan_err,
+};
+use datafusion_expr::Expr;
+use datafusion_expr::execution_props::ExecutionProps;
+use datafusion_expr::simplify::SimplifyContext;
+use std::sync::Arc;
+
+/// Parse and simplifies a SQL expression to a numeric literal,
+/// corresponding to an arrow primitive type `T` (for example, Float64Type).
+///
+/// This function simplifies and coerces the expression, then extracts the underlying
+/// native type using `TryFrom<ScalarValue>`.
+///
+/// # Example
+/// ```ignore
+/// let value: f64 = parse_sql_literal::<Float64Type>(expr)?;
+/// ```
+pub fn parse_sql_literal<T>(expr: &Expr) -> Result<T::Native>
+where
+    T: ArrowPrimitiveType,
+    T::Native: TryFrom<ScalarValue, Error = DataFusionError>,
+{
+    // Empty schema is sufficient because it parses only literal expressions
+    let schema = DFSchemaRef::new(DFSchema::empty());
+
+    log::debug!("Parsing expr {:?} to type {}", expr, T::DATA_TYPE);
+
+    let execution_props = ExecutionProps::new();
+    let simplifier = ExprSimplifier::new(
+        SimplifyContext::new(&execution_props).with_schema(Arc::clone(&schema)),
+    );
+
+    // Simplify and coerce expression in case of constant arithmetic operations (e.g., 10 + 5)
+    let simplified_expr: Expr = simplifier
+        .simplify(expr.clone())
+        .map_err(|err| plan_datafusion_err!("Cannot simplify {expr:?}: {err}"))?;
+    let coerced_expr: Expr = simplifier.coerce(simplified_expr, schema.as_ref())?;
+    log::debug!("Coerced expression: {:?}", &coerced_expr);
+
+    match coerced_expr {
+        Expr::Literal(scalar_value, _) => {
+            // It is a literal - proceed to the underlying value
+            // Cast to the target type if needed
+            let casted_scalar = scalar_value.cast_to(&T::DATA_TYPE)?;
+
+            // Extract the native type
+            T::Native::try_from(casted_scalar).map_err(|err| {
+                plan_datafusion_err!(
+                    "Cannot extract {} from scalar value: {err}",
+                    std::any::type_name::<T>()
+                )
+            })
+        }
+        actual => {
+            plan_err!(
+                "Cannot extract literal from coerced {actual:?} expression given {expr:?} expression"
+            )
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use arrow::datatypes::{DataType, Float64Type, Int64Type};
+    use datafusion_common::config::ConfigOptions;
+    use datafusion_common::{TableReference, not_impl_err};
+    use datafusion_expr::planner::{ContextProvider, RelationPlannerContext};
+    use datafusion_expr::sqlparser::parser::Parser;
+    use datafusion_expr::{AggregateUDF, ScalarUDF, TableSource, WindowUDF};
+    use datafusion_sql::planner::{PlannerContext, SqlToRel};
+    use datafusion_sql::relation::SqlToRelRelationContext;
+    use datafusion_sql::sqlparser::dialect::GenericDialect;
+    use std::sync::Arc;
+
+    // Simple mock context provider for testing
+    struct MockContextProvider {
```

Comment:
These tests can probably be simplified to remove the SQL parsing related code and only supply Expr to test cases